chore: adds quality gate for rerunning e2e spec files that are new or have been modified

seaona opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**
This PR adds a quality gate for new or modified e2e spec files. Whenever there is a PR which modifies or changes a test, this will be run more times, in order to prevent introducing a flakiness accidentally. It is done as follows:


- Identifies any new or modified e2e file from inside the test/ folder using `git diff` and using these 2 filters:
  - `file.filename.startsWith('test/e2e/') &&`
  - `file.filename.endsWith('.spec.js') || file.filename.endsWith('.spec.ts')  `
- Copies the given specs x5 times in the list of testpaths to execute -> this number is arbitrary, we could modify it to any value we want. The reason for taking this approach instead of changing the retrial number is to benefit of the parallelization, as @HowardBraham pointed out in a comment.
- Since we already had a flag which could support the re-running successful tests, `--retry-until-failure` I just leveraged this into the `for` loop for each test, and if that testcase was identified as new/modified, the flag is added so the new tests fail fast without retrials

### Incremental git fetch depth within shallow clone
We use git fetch with incremental depth as @danjm  suggested. The ci environment uses a shallow clone, meaning we won't be able to succeed just by using git diff as it won't find the merge base. For fixing that, we start with a git fetch depth of 1, and keep incrementing the depth (1, 10, 100) it the error is `no merge base` up until 100. If the git diff still fails, we then do a full git fetch with the `unshallow` flag.

- [Update] This is the working result with the last commit https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/89269/workflows/103b78a8-8f0d-4323-96b0-9e235c4cbc81/jobs/3296802
![Screenshot from 2024-06-26 11-39-19](https://github.com/MetaMask/metamask-extension/assets/54408225/a2a89d6a-3a73-48ba-91a3-20aeadc38573)


### New ci Job
The git diff is done in a new ci job which runs at the beginning in parallel of prep-deps.





[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24556?quickstart=1)


## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/24009

## **Manual testing steps**

1. Check ci runs (notice previous runs had failing and changed tests on purpose, in order to try the different scenarios described below)

## **Screenshots/Recordings**


=============================================== [UPDATE with the new code changes]
- :green_circle:  Case 1: A test has changed -> it's rerun 1+5 times and it's successful (it will be run in different buckets)



https://github.com/MetaMask/metamask-extension/assets/54408225/c1456104-1f5f-4ef3-9364-4e435f8797f4


https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/89277/workflows/7fce0a2e-773f-46da-8ab9-1dbec7992b58/jobs/3297267/parallel-runs/10?filterBy=ALL


- :green_circle:  Case 2: A test has changed, but it has a mistake in the code (intentionally to simulate a flaky test) -> it fails immediately and there are no more retries. The rest of the tests, are retried if they failed as usual


  - Case for main test build test: https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/89277/workflows/7fce0a2e-773f-46da-8ab9-1dbec7992b58/jobs/3297267/artifacts
  - Case for flask specific test: https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/89277/workflows/7fce0a2e-773f-46da-8ab9-1dbec7992b58/jobs/3297277/artifacts



- :green_circle:  Case 3: A PR has no test spec files changed -> nothing different happens
- ci run: check current ci 


## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

seaona committed:
quality gate mock alt

seaona commented on test/e2e/run-all.js:
```diff
@@ -212,12 +213,19 @@ async function main() {
+      const testFileName = testPath.split('/').pop();
+      const isChangedOrNew = changedOrNewTests.includes(testFileName);
+      const extraArgs = isChangedOrNew
+        ? ['--retry-until-failure=true', '--retries=5']
```
the number 5 is arbitrary, we could decide how many times we want to run them

seaona committed:
fix export

seaona committed:
fix PR number match

seaona committed:
overwrite retries number for retry-until-failure cases

seaona committed:
change file to ts and add log for debugging

seaona committed:
revert js

seaona commented on test/e2e/run-all.js:
```diff
@@ -212,12 +213,26 @@ async function main() {
   console.log('My test list:', myTestList);
 
+  const changedOrNewTests = await fetchChangedE2eFiles();
+  console.log('Spec files that will be re-run:', changedOrNewTests);
```
this is left for debugging purposes

HowardBraham replied:
You're going to sacrifice some parallel execution speed by doing it this way.  It also only has to be done if it's running on CircleCI, and not locally.

It would be a better approach if at the top of `function runningOnCircleCI()`, you looked at `testPaths` and then duplicated 5 times each thing that was also in `changedOrNewTests`.

This would allow it to distribute the workload evenly across the VMs.

seaona replied:
thank you for your suggestion @HowardBraham That's definetly a good point! I was thinking it would be negligible given the small amount of retries, but it's true that we can benefit from parallelization witha small tweak. I'll add the changes :bowing_woman:

seaona replied:
changes added :+1:

seaona commented on test/e2e/run-all.js:
```diff
@@ -212,12 +213,26 @@ async function main() {
+      }
+
+      const extraArgs = isTestChangedOrNew
+        ? ['--retry-until-failure', '--retries=5']
```
the number 5 is arbitrary, we could change it if another number is better suited

DDDDDanica replied:
Shall we extract it to a const and then it is easier to make modifications in the future

seaona replied:
thank you for your suggestion @DDDDDanica :bowing_woman: I've moved this into a const :+1:

seaona committed:
add ts spec files to the filter and make another test fail to verify logic is working properly in failing tests

DDDDDanica commented on test/e2e/run-all.js:
```diff
@@ -212,12 +213,31 @@ async function main() {
+  let changedOrNewTests = '';
+  try {
+    changedOrNewTests = await fetchChangedE2eFiles();
+  } catch (error) {
+    console.error('Error fetching changed e2e files:', error);
+  }
```
SUggestion: we can remove the catch here as it is handled in `fetchChangedE2eFiles`

seaona replied:
thank you for your suggestion. I've updated the `fetchChangedE2eFiles` so in the case something goes wrong it returns an empty string. This way we can remove this try/catch so we are always safe

DDDDDanica replied:
Thank you !

seaona committed:
address dev review: move retries to variable remove try/catch

seaona committed:
leave the specs as they were before (changed for ci testing purposes)

seaona commented on development/lib/retry.js:
```diff
@@ -46,18 +46,22 @@ async function retry(
+  if (retryUntilFailure) {
+    return null;
+  }
+
```
before it assumed that if we reach the max retries, it was an error and throw the error with rejection message `'Retry limit reached'`, but in the case of using retryUntilFailure, reaching the max retry limit it's actually a good thing, as it means the test has not failed. That's why in this case I'm returning just null

HowardBraham replied:
Can we rename this parameter `stopAfterOneFailure` for better clarity?

seaona replied:
changed :+1:

seaona committed:
Merge branch 'develop' into quality-gate-gh

danjm commented on test/e2e/fetch-changed-files.js:
```diff
@@ -0,0 +1,33 @@
+const axios = require('axios');
```
I'm not sure, but we might be able to avoid a fetch to github here, and instead just use git. Not sure if we have the repo and git history in the right time and place on CI to do it, but it would be good if we could: more efficient than a network request, and not dependent on network conditions or the github api possibly being down.

seaona replied:
thank you Dan! That is a good point. I have an alternative branch where I explored a bit the git ci option, however it was not straight-forward and I believe we might need to tweak some things on the config.yml file in order to accomplish it. I decided to go for this option to not over-engineer it, but for the reason you mention it might be worth to try the ci option a bit further.



https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/80853/workflows/9ac00125-66fc-400d-82ac-3f22b4c928c1/jobs/2859112

legobeat replied:
I didn't try but something like this might work?

``
gitDiff = await git.diffSummary(['--name-only',  `origin/develop...${process.env.CIRCLE_SHA1}`, 'test/**/*.spec.js']);
```

https://github.com/MetaMask/metamask-extension/blob/develop/development/generate-rc-commits.js#L2

https://github.com/steveukx/git-js?tab=readme-ov-file#git-diff

HowardBraham replied:
Does the local git command like that work if you're doing a shallow checkout?  Also, the way this is written, it will be hitting the GitHub API hundreds of times per workflow (because there are hundreds of parallel machines running the workflow).  Perhaps do it in `prep-deps`, and then `persist_to_workspace`?

legobeat replied:
Good point on the shallow checkout. Could be addressed by a `git fetch --negotiation-tip=develop origin e23f55d80e2c97d56f4fd929873b87e7f465b351` (possibly in `prep-deps`)?

seaona replied:
thank you for the suggestions :pray: :heart:  In the shallow clone we do this:

```
 git clone --depth 1 --no-checkout "$CIRCLE_REPOSITORY_URL" .
 git fetch --depth 1 origin "$CIRCLE_SHA1"  
```
I was thinking we would need to do git fetch develop explicitly to get available the develop branch commits in the ci environment? :thinking:  I'm also not sure if we would need to increase the depth too here :thinking: I can investigate this further

seaona replied:
logic for git fetch with depth and git diff has been added

seaona replied:
closing as this is now resolved :+1:

HowardBraham commented on test/e2e/run-all.js:
```diff
@@ -212,12 +213,28 @@ async function main() {
+
+      const testFileName = testPath.split('/').pop();
+      const isTestChangedOrNew = changedOrNewTests?.includes(testFileName);
+      const retryIndex = args.indexOf('--retries');
+      if (retryIndex !== -1) {
+        args.splice(retryIndex, 2);
+      }
+
+      const extraArgs = isTestChangedOrNew
+        ? ['--retry-until-failure', `--retries=${retriesForChangedOrNewTests}`]
+        : [];
```
There's no need for this duct-taped workaround complexity.  Look at line 191 in this file.  Just put `retries` into `extraArgs` down here instead of `args` up there.

HowardBraham replied:
(if you go with my other suggestion about `runningOnCircleCI` though, you can probably remove most of this anyway)

seaona replied:
thank you!! if we follow the approach of copying the spec files x5 times on `runningOnCircleCI` then we can indeed remove all of this logic. Just something I'm wondering, would it be okay then to treat each of those new specs the same as rest (same flags)? or we still want them to fail immediately with no-retries?

Given that @legobeat has a PR for reducing test retries, maybe it's fine to treat all the tests equally; then there's no need to add any extra flag, and just copy the spec file x times in the test list from runningOnCiircleCi, and this would simplify a lot the logic

What do you think?

seaona replied:
:information_source:  retries has been removed and instead we add them in the testpaths x5 times, so we benefit from parallelization.
Now only the argument for failing fast is passed here

seaona committed:
Quality gate gh ci job (#24787)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24787?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

danjm commented on .circleci/scripts/get-changed-files.sh:
```diff
@@ -0,0 +1,21 @@
+echo "$DIFF_RESULT"
+
+# Store the output of git diff
+git diff --name-only develop..."$CIRCLE_SHA1" >> changed-files/changed-files.txt
```
maybe this should be a diff with `origin/develop`? because above there is the code to `git fetch origin develop`, but those commits are not pulled locally

seaona replied:
thank you for the suggestion, I tried it but it does not seem to make any difference :thinking: 


![Screenshot from 2024-06-04 11-53-08](https://github.com/MetaMask/metamask-extension/assets/54408225/1e50850a-76d5-4a19-929a-4177a1ca31bb)

seaona replied:
:information_source: the problem was that the branch was way behind the depth. Updating the branch fixed it

seaona committed:
address comments

seaona committed:
origin develop

seaona committed:
depth update

seaona committed:
50 fetch

seaona committed:
rename funcs

seaona committed:
Merge branch 'develop' into quality-gate-gh

seaona committed:
switch from sh to js file

seaona committed:
fix changedfilesUtils

seaona committed:
move script to config file

seaona committed:
git diff incremental depth

seaona committed:
fix git fetch incremental

seaona committed:
git fetch pr branch with depth too

seaona committed:
console log path

seaona committed:
define output path

seaona committed:
format spec entries with path and new lines

seaona committed:
fix back the spec files for testing ci

seaona committed:
only do for loop if there are changed e2e files

seaona commented on .circleci/scripts/git-diff-develop.js:
```diff
@@ -0,0 +1,68 @@
+}
+
+async function gitDiffWithRetry() {
+  const depths = [5, 10, 15, 20, 30, 40, 50];
```
to consume fewer resources, we start with small depth jumps +5 and later on we increment with +10 jumps

seaona commented on .circleci/scripts/git-diff-develop.js:
```diff
@@ -0,0 +1,68 @@
+    console.error('An error occurred:', error.message);
+    process.exit(1);
+  }
+}
```
leaving all the logs so we are able to debug ci if needed

chloeYue commented:
I have reviewed the latest version and LGTM. I think it would be worthwhile to tag @danjm, @HowardBraham, and @legobeat to give another review as you have commented. Let's try to proceed with this PR.

hjetpoluru reviewed (dismissed):
LGTM

seaona committed:
update git diff fallback and adjust
depth intervals

seaona committed:
testing git diff with full checkout

seaona committed:
fix fetch develop

seaona committed:
unshallow

seaona commented on .circleci/scripts/git-diff-develop.js:
```diff
@@ -0,0 +1,76 @@
+
+    const { stdout: diffResult } = await exec(`git diff --name-only origin/develop...${process.env.CIRCLE_BRANCH}`);
+    diffOutput = diffResult;
+
```
this new update has been tested and works correctly - check the logs [here](https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/87981/workflows/4095b4ea-f19d-498b-8aed-7a694670ba68/jobs/3230260)




![Screenshot from 2024-06-18 18-49-23](https://github.com/MetaMask/metamask-extension/assets/54408225/7bff2d87-97f8-4d96-bc00-b83f8c94c4c0)

seaona committed:
separate job

seaona committed:
remove requried job

seaona committed:
Merge branch 'develop' into quality-gate-gh

seaona committed:
Merge branch 'develop' into quality-gate-gh

Gudahtt commented on .circleci/scripts/git-diff-develop.js:
Nit: we could consider writing this in TypeScript, and running it with `tsx`

seaona replied:
thank you! changed to `ts`  :+1:

Gudahtt commented on .circleci/scripts/git-diff-develop.js:
```diff
@@ -0,0 +1,76 @@
+const path = require('path');
+
+
+async function fetchWithDepthIncrement(depth) {
```
Nit: It would be create if these had JSDoc comments

seaona replied:
documentation added :+1:

Gudahtt commented on .circleci/scripts/git-diff-develop.js:
```diff
@@ -0,0 +1,76 @@
+      console.log(`Attempting git diff with depth ${depth}...`);
+      await fetchWithDepthIncrement(depth);
+
+      const { stdout: diffResult } = await exec(`git diff --name-only origin/develop...${process.env.CIRCLE_BRANCH}`);
```
Nit: In each of these places where we use `origin/develop`, perhaps we could use `origin/HEAD` instead? That should represent the default branch, and will keep working if/when we change it from `develop`.

seaona replied:
great suggestion! Changed :+1:

Gudahtt commented on .circleci/scripts/git-diff-develop.js:
```diff
@@ -0,0 +1,76 @@
+      console.log(`Attempting git diff with depth ${depth}...`);
+      await fetchWithDepthIncrement(depth);
+
+      const { stdout: diffResult } = await exec(`git diff --name-only origin/develop...${process.env.CIRCLE_BRANCH}`);
```
Nit: Perhaps we could do this in two parts - find the merge base, then diff once we have it. This may be faster, and would avoid having to parse strings to determine if we found what we needed.

This command can tell us what the merge-base is: `git merge-base origin/HEAD HEAD` (we use it elsewhere in CI)
And if it can't be found, it returns with a non-zero exit code. 

e.g.

```typescript
function fetchUntilMergeBaseFound() {
  const depths = [1, 10, 100];
  for (const depth of depths) {
    console.log(`Attempting git diff with depth ${depth}...`);
    await fetchWithDepthIncrement(depth);

    try {
      await exec(`git merge-base origin/HEAD HEAD`);
      return;
    } catch (error) {
      // `git merge-base` returns an exit code of 1 when no match is found
      if (error.code === 1) {
        console.error(`Error 'no merge base' encountered with depth ${depth}. Incrementing depth...`);
      } else {
        throw error;
      }
    }
  }
  await exec(`git fetch --unshallow origin develop`);
}
```

Then we can call this once and get the diff, without repeating the diff command multiple times

seaona replied:
thank you for the suggestion, Mark. I've split it up in two functions now

Gudahtt commented on test/e2e/changedFilesUtil.js:
```diff
@@ -0,0 +1,30 @@
+const fs = require('fs').promises;
+
+const BASE_PATH = '/home/circleci/project/';
```
Nit: perhaps we could make this more portable by using `path.reoslve(__dirname, '..', '..')` instead

seaona replied:
changes added :+1:

Gudahtt commented on test/e2e/changedFilesUtil.js:
```diff
@@ -0,0 +1,30 @@
+const BASE_PATH = '/home/circleci/project/';
+const CHANGED_FILES_PATH = `${BASE_PATH}changed-files/changed-files.txt`;
+
+async function readChangedFiles() {
```
Nit: Comments would be great for these two as well!

Gudahtt commented on test/e2e/run-all.js:
```diff
@@ -217,7 +231,13 @@ async function main() {
       console.log(`\nExecuting testPath: ${testPath}\n`);
-      await runInShell('node', [...args, testPath]);
+      const testFileName = testPath.split('/').pop();
+      const isTestChangedOrNew = changedOrNewTests?.includes(testFileName);
```
This may produce strange results if this test filename matches a subset of another test filename

Instead perhaps we can return an array from `filterE2eChangedFiles` rather than a string, and check here that this matches a filename exactly.

seaona replied:
thank you for the suggestion! I'm now returning an array from the filtereE2eChangedFiles function :+1:

Gudahtt commented:
Looks great! I can take another look when this is out of draft

seaona committed:
Merge branch 'develop' into quality-gate-gh

seaona committed:
addressing some of the review comments

seaona committed:
change .js to .ts

seaona committed:
addressed comments and test failing test

seaona committed:
update function name

seaona committed:
add console logs and remove extra path

seaona committed:
tweak path

seaona committed:
add missing /

seaona committed:
remove changes used to test ci

seaona committed:
cleanup

chloeYue reviewed (dismissed):
I took another look on the latest version, lots of improvements compared to the first version, congrats!

Gudahtt commented on .circleci/scripts/git-diff-develop.ts:
```diff
@@ -0,0 +1,94 @@
+    await exec(`git fetch --depth ${depth} origin develop`);
+    await exec(`git fetch --depth ${depth} origin ${process.env.CIRCLE_BRANCH}`);
+    return true;
+  } catch (error: any) {
```
Nit: This should be disallowed by our lint rules, but our configuration is currently broken and is not applying TypeScript lint rules to this file.

We can avoid using `any` here by using `unknown` instead. TypeScript will then complain about us using `error.message`, but this is easily resolved with a type check (e.g. `error instanceof Error`), or by passing `error` directly into `console.error (e.g. `console.error(`Failed to fetch with depth ${depth}:`, error);`)

seaona replied:
fixed

Gudahtt commented on .circleci/scripts/git-diff-develop.ts:
```diff
@@ -0,0 +1,94 @@
+    try {
+      await exec(`git merge-base origin/HEAD HEAD`);
+      return;
+    } catch (error: any) {
+      if (error.code === 1) {
```
Nit: This is what I would use to replace `any` here:
```suggestion
    } catch (error: unknown) {
      if (
        error instanceof Error &&
        hasProperty(error, 'code') &&
        error.code === 1
      ) {
```

Gudahtt commented on .circleci/scripts/git-diff-develop.ts:
```diff
@@ -0,0 +1,94 @@
+/**
+ * Fetches the git repository with a specified depth.
+ *
+ * @param {number} depth - The depth to use for the fetch command.
```
Nit: We should omit the types from these comments in TypeScript files, since the types are already documented in the type signature. They can make the comments overly verbose, and can easily get out of sync.

e.g.
```suggestion
 * @param depth - The depth to use for the fetch command.
```

Technically for TypeScript files we're following something closer to the TSDoc standard: https://tsdoc.org
It's slightly different from JSDoc, it has fewer tags, and some extra ones specific to TypeScript.

seaona replied:
ooh thank you for the detailed explanation Mark :heart: make sense to me

Gudahtt commented on .circleci/scripts/git-diff-develop.ts:
```diff
@@ -0,0 +1,94 @@
+  }
+}
+
+storeGitDiffOutput();
```
Nit: This is something our linter should be catching as well (missing newline) 
```suggestion
storeGitDiffOutput();

```

Gudahtt commented on test/e2e/run-all.js:
```diff
@@ -30,9 +31,44 @@ const getTestPathsForTestDir = async (testDir) => {
+ */
+async function applyQualityGate(fullTestList) {
+  let qualityGatedList = fullTestList;
+  changedOrNewTests = await filterE2eChangedFiles();
```
Nit: Perhaps we could generate this earlier and pass it in, or return it alongside the quality gate list?

With this variable at the module scope, it may be easy to introduce errors later if we try to use it before `applyQualityGate` is called. And if/when we convert this file to TypeScript, it will make the type `| undefined` which is more difficult to deal with. We can avoid both problems by using `cost` to declare the variable, either in `runningOnCircleCI` or in both functions.

seaona replied:
I've declared it in `runningOnCirclCI` and then passed the `changedOrNewTests` it to `applyQualityGate`. I'm also returning this on the `runningOnCircleCi` since we need it to apply the no-retry flag for the tests that are new/changed

Gudahtt commented on test/e2e/run-all.js:
```diff
@@ -30,9 +31,44 @@ const getTestPathsForTestDir = async (testDir) => {
-function runningOnCircleCI(testPaths) {
-  const fullTestList = testPaths.join('\n');
+async function runningOnCircleCI(testPaths) {
+  let fullTestList = testPaths.join('\n');
+  fullTestList = await applyQualityGate(fullTestList);
```
Nit: Slight simplification, lets us use `const` which makes tracing this code a bit easier.
```suggestion
  const fullTestList = await applyQualityGate(testPaths.join('\n'));
```

Or alternatively:

```suggestion
  const testPathsList = testPaths.join('\n');
  const fullTestList = await applyQualityGate(testPathsList);
```

Gudahtt commented on .circleci/config.yml:
```diff
@@ -472,6 +484,21 @@ jobs:
             - node_modules
             - build-artifacts
 
+  get-changed-files-with-git-diff:
```
Nit: I see that this is now expected to be run before `run-all.js`. But this dependency may not be clear for readers. perhaps we could add a comment here, or somewhere, to explain that this must be a requirement of all E2E test runs

Gudahtt reviewed (dismissed):
LGTM!

seaona committed:
add missing newline Mark's suggestion

Co-authored-by: Mark Stacey <markjstacey@gmail.com>

seaona committed:
Omit type from tsdoc - Mark's suggestion

Co-authored-by: Mark Stacey <markjstacey@gmail.com>

seaona committed:
simplifying variables - Mark's suggestion

Co-authored-by: Mark Stacey <markjstacey@gmail.com>

seaona committed:
avoid using any in ts - Mark's suggestion

Co-authored-by: Mark Stacey <markjstacey@gmail.com>

seaona committed:
add job comment for config

seaona committed:
changed changedOrnew var definition

Gudahtt commented on .circleci/scripts/git-diff-develop.ts:
```diff
@@ -0,0 +1,99 @@
+    await fetchWithDepth(depth);
+
+    try {
+      await exec(`git merge-base origin/HEAD HEAD`);
+      return;
+    } catch (error: unknown) {
+      if (
+        error instanceof Error &&
+        hasProperty(error, 'code') &&
+        error.code === 1
+      ) {
+        console.error(`Error 'no merge base' encountered with depth ${depth}. Incrementing depth...`);
+      } else {
+        throw error;
+      }
+    }
+  }
+  await exec(`git fetch --unshallow origin develop`);
+}
+
+/**
+ * Performs a git diff command to get the list of files changed between the current branch and the origin.
+ * It first ensures that the necessary commits are fetched until the merge base is found.
+ *
+ * @returns {Promise<string>} The output of the git diff command, listing the changed files.
+ * @throws {Error} If unable to get the diff after fetching the merge base or if an unexpected error occurs.
+ */
+async function gitDiff(): Promise<string> {
+  await fetchUntilMergeBaseFound();
+  const { stdout: diffResult } = await exec(`git diff --name-only origin/HEAD...${process.env.CIRCLE_BRANCH}`);
+  if (!diffResult) {
+      throw new Error('Unable to get diff after full checkout.');
+  }
+  return diffResult;
+}
+
+/**
+ * Stores the output of git diff to a file.
+ *
+ * @returns {Promise<void>} Returns a promise that resolves when the git diff output is successfully stored.
```
Nit: The types are still present in the rest of the doc comments. This suggestion should remove them:

```suggestion
 * @returns True if the fetch is successful, otherwise false.
 */
async function fetchWithDepth(depth: number): Promise<boolean> {
  try {
    await exec(`git fetch --depth ${depth} origin develop`);
    await exec(`git fetch --depth ${depth} origin ${process.env.CIRCLE_BRANCH}`);
    return true;
  } catch (error: unknown) {
    console.error(`Failed to fetch with depth ${depth}:`, error);
    return false;
  }
}

/**
 * Attempts to fetch the necessary commits until the merge base is found.
 * It tries different fetch depths and performs a full fetch if needed.
 *
 * @throws If an unexpected error occurs during the execution of git commands.
 */
async function fetchUntilMergeBaseFound() {
  const depths = [1, 10, 100];
  for (const depth of depths) {
    console.log(`Attempting git diff with depth ${depth}...`);
    await fetchWithDepth(depth);

    try {
      await exec(`git merge-base origin/HEAD HEAD`);
      return;
    } catch (error: unknown) {
      if (
        error instanceof Error &&
        hasProperty(error, 'code') &&
        error.code === 1
      ) {
        console.error(`Error 'no merge base' encountered with depth ${depth}. Incrementing depth...`);
      } else {
        throw error;
      }
    }
  }
  await exec(`git fetch --unshallow origin develop`);
}

/**
 * Performs a git diff command to get the list of files changed between the current branch and the origin.
 * It first ensures that the necessary commits are fetched until the merge base is found.
 *
 * @returns The output of the git diff command, listing the changed files.
 * @throws If unable to get the diff after fetching the merge base or if an unexpected error occurs.
 */
async function gitDiff(): Promise<string> {
  await fetchUntilMergeBaseFound();
  const { stdout: diffResult } = await exec(`git diff --name-only origin/HEAD...${process.env.CIRCLE_BRANCH}`);
  if (!diffResult) {
      throw new Error('Unable to get diff after full checkout.');
  }
  return diffResult;
}

/**
 * Stores the output of git diff to a file.
 *
 * @returns Returns a promise that resolves when the git diff output is successfully stored.
```

seaona committed:
missing to remove the rest of the types in the descriptions - Mark's suggestion

Co-authored-by: Mark Stacey <markjstacey@gmail.com>

Gudahtt reviewed (approved):
LGTM!

HowardBraham committed:
Merge branch 'develop' into quality-gate-gh
